File: assets/Tools/map_from_bitmap.py
from PIL import Image
import os
from random import randint
import argparse

parser = argparse.ArgumentParser("map_from_bitmap")
parser.add_argument(
    "filename", help="the name of the input image excluding .bmp")
parser.add_argument("-s", "--spawns", nargs="+", type=int,
                    default=[0], help="Create random spawns. default is 0")
args = parser.parse_args()
filename = args.filename
spawns = tuple(args.spawns) if len(
    args.spawns) == 2 else tuple(args.spawns * 2)
assert len(spawns) == 2, spawns


class MapFromBitmap:
    """Create a .txt map file from a .bmp bitmap file
    The .bmp must be in the /assets/Tools/bitmaps/ folder
    Run the file from /assets/Tools/
    The .txt is called output.txt and is placed in /assets/Tools/

    Params:
    str filename_: the name of the .bmp excluding .bmp
    tuple[int] num_spawns: a tuple where the first is the number of
        zombie spawns and the second is pickup spawns

    Raises:
    RuntimeError if .create() is run more than once

    """
    blacks = {0, (0, 0, 0), (0, 0, 0, 0)}  # Some files hsave different values
    # for black. This is monochrome, RGB, and RGBA respectivly
    cwd = os.getcwd()  # current working directory

    def __init__(self, filename_, num_spawns=(5, 5)):
        img_path = MapFromBitmap.cwd + "/bitmaps/{}.bmp".format(filename_)
        self.img = Image.open(img_path)
        self.output_path = MapFromBitmap.cwd + "/output.txt"
        self.new_file = open(self.output_path, mode="w")
        self.width, self.height = self.img.size
        self.num_spawns = num_spawns
        assert sum(num_spawns) <= self.width * self.height
        self.z_spawns, self.p_spawns = self.get_spawns()
        self.as_str = ""

    @classmethod
    def test(cls, num_spawns=(2, 3)):
        return cls("Test", num_spawns)

    def get_spawns(self):
        taken_spawns = []
        for num_spawn in self.num_spawns:
            current_spawns = []
            while len(current_spawns) < num_spawn:
                rnd_num = randint(0, self.width * self.height - 1)
                coordinate = divmod(rnd_num, self.height)
                if (not is_black(self.img.getpixel(coordinate)) and
                        coordinate not in taken_spawns):
                    current_spawns.append(coordinate)
                    taken_spawns.append(coordinate)
            yield current_spawns

    def create_file(self):
        if self.as_str:
            raise RuntimeError("Already created file")
        for i in range(self.height * self.width):
            y, x = coordinate = divmod(i, self.width)
            pixel = self.img.getpixel(coordinate)
            tile = "#" if is_black(pixel) else "."
            if coordinate in self.z_spawns:
                tile = "Z"
            elif coordinate in self.p_spawns:
                tile = "P"
            self.new_file.write(tile)
            if (i + 1) % self.width == 0:
                self.new_file.write("\n")

        self.new_file.close()
        self.img.close()
        with open(self.output_path, mode="r") as file:
            self.as_str = file.read()
        if self.num_spawns == (0, 0):
            print("don't forget to add zombie and pickup spawn points!")

    def display(self):
        print(self.as_str)


def is_black(pixel) -> bool:
    return pixel in MapFromBitmap.blacks


if __name__ == "__main__":
    a = MapFromBitmap(filename, spawns)
    a.create_file()
    a.display()

File: assets/Tools/MapFromBitmap.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Drawing;
using System.IO;

namespace MapFromBitmap
{
    class Program
    {
        static void Main(string[] args)
        {
            string cwd = Environment.CurrentDirectory + "\\";
            string filename;
            try
            {
                filename = args[0];
            }
            catch (IndexOutOfRangeException)
            {
                filename = "test";
            }
            string path = cwd + "bitmaps\\" + filename + ".bmp";
            Bitmap bmp = new Bitmap(path);
            int[] nums = GetNums(args);
            var spawns = GetSpawns(bmp, nums);
            var z_spawns = spawns[0];
            var p_spawns = spawns[1];
            List<string> map = new List<string> { };
            for (int i = 0; i < bmp.Height * bmp.Width; i++)
            {
                int x = i % bmp.Width;
                int y = i / bmp.Width;
                string tile = "";
                var pixel = bmp.GetPixel(x, y);
                if (IsBlack(pixel))
                    tile = "#";
                else
                    tile = ".";
                int[] coordinate = new int[] { x, y };

                if (IsArrayInArrayOfArrays(z_spawns, coordinate))
                    tile = "Z";
                else if (IsArrayInArrayOfArrays(p_spawns, coordinate))
                    tile = "P";

                map.Add(tile);
                if ((x + 1) % bmp.Width == 0 && y != bmp.Height - 1)
                {
                    map.Add(Environment.NewLine);
                }
            }
            string AsString = string.Join("", map);
            Console.WriteLine(AsString);
            File.WriteAllText(cwd + "output.txt", AsString);
            }
        static List<int[]>[] GetSpawns(Bitmap bmp, int[] nums)
        {
            var return_val = new List<int[]>[2];
            List<int[]> taken = new List<int[]> { };
            Random random_gen = new Random();
            for(int i = 0; i < 2; i++)
            {
                List<int[]> current_spawns = new List<int[]> { };
                while (current_spawns.Count() < nums[i])
                {
                    int rnd = random_gen.Next(0, bmp.Width * bmp.Height);
                    int x = rnd / bmp.Height;
                    int y = rnd % bmp.Height;
                    int[] coordinate = new int[] { x, y };
                    var pixel = bmp.GetPixel( x, y );
                    if ((!IsBlack(pixel)) && (!IsArrayInArrayOfArrays(taken, coordinate)))
                    {
                        current_spawns.Add(coordinate);
                        taken.Add(coordinate);
                    }
                }
                return_val[i] = current_spawns;
            }
            return return_val;
        }

        static int[] GetNums(string[] args)
        {
            int[] nums = new int[2];

            if (args.Count() <= 1 || args.Count() >= 4)
            {
                raise new ArgumentOutOfRangeException("Too many or too few arguments");
            }
            else if (args.Count() == 2)
            {
                nums[0] = int.Parse(args[1]);
                nums[1] = nums[0];
            }
            else if (args.Count() == 3)
            {
                nums[0] = int.Parse(args[1]);
                nums[1] = int.Parse(args[2]);
            }
            else
            {
                throw new ArgumentOutOfRangeException("Too many command line arguments");
            }
            return nums;
        }

        static bool IsBlack(Color pixel)
        {
            return pixel.R == 0 && pixel.G == 0 && pixel.B == 0;
        }

        static bool IsArrayInArrayOfArrays<T>(IList<T[]> array1, IList<T> array2)
        {
            foreach(var element in array1)
            {
                if (element.SequenceEqual(array2))
                    return true;
            }
            return false;
        }
    }
}
